File: crates/ferrogate-observability/src/cloudflare_test.rs
// Token4AI Cloud Attribution
// Developed by the commercial cloud service company represented by https://token4ai.cloud.
// Author: jamesduan (X: https://x.com/JamesDuanL)
// Created: 2026-07-26
// description: Token4AI Cloud, FerroGate AI Gateway, Rust API Gateway, agent-native AI traffic infrastructure.

use super::*;
use crate::otlp::OtlpAttribute;

const TOKEN: &str = "s3cr3t-collector-token";
const ENDPOINT: &str = "https://telemetry-collector.example.workers.dev";

fn backend() -> CloudflareBackend {
    CloudflareBackend::new(ENDPOINT, TOKEN)
}

fn snapshot() -> GatewayMetricsSnapshot {
    GatewayMetricsSnapshot {
        service_name: "ferrogate".to_string(),
        ..Default::default()
    }
}

fn span() -> OtlpSpanRecord {
    OtlpSpanRecord {
        trace_id: "0af7651916cd43dd8448eb211c80319c".to_string(),
        span_id: "b7ad6b7169203331".to_string(),
        parent_span_id: None,
        name: "ferrogate.gateway.request".to_string(),
        start_time_unix_nano: 1,
        end_time_unix_nano: 2,
        attributes: vec![OtlpAttribute::new("tenant", "acme")],
    }
}

fn log() -> OtlpLogRecord {
    OtlpLogRecord {
        trace_id: None,
        span_id: None,
        severity_text: "INFO".to_string(),
        body: "request".to_string(),
        time_unix_nano: 1,
        attributes: Vec::new(),
    }
}

fn header<'a>(request: &'a OtlpHttpRequest, name: &str) -> Option<&'a str> {
    request
        .headers
        .iter()
        .find(|(key, _)| key.eq_ignore_ascii_case(name))
        .map(|(_, value)| value.as_str())
}

#[test]
fn every_signal_carries_the_bearer_credential() {
    let backend = backend();

    let requests = [
        backend.metrics_request(&snapshot()).unwrap().unwrap(),
        backend
            .traces_request("ferrogate", &[span()])
            .unwrap()
            .unwrap(),
        backend
            .logs_request("ferrogate", &[log()])
            .unwrap()
            .unwrap(),
    ];

    for request in &requests {
        assert_eq!(
            header(request, "authorization"),
            Some(format!("Bearer {TOKEN}").as_str())
        );
    }
}

#[test]
fn signals_target_the_standard_otlp_paths() {
    let backend = backend();

    assert_eq!(
        backend.metrics_request(&snapshot()).unwrap().unwrap().url,
        format!("{ENDPOINT}/v1/metrics")
    );
    assert_eq!(
        backend
            .traces_request("ferrogate", &[span()])
            .unwrap()
            .unwrap()
            .url,
        format!("{ENDPOINT}/v1/traces")
    );
    assert_eq!(
        backend
            .logs_request("ferrogate", &[log()])
            .unwrap()
            .unwrap()
            .url,
        format!("{ENDPOINT}/v1/logs")
    );
}

#[test]
fn debug_output_redacts_the_credential() {
    // The export loop debug-logs its sink at startup; a derived `Debug` would
    // put the collector token in the process log.
    let rendered = format!("{:?}", backend());
    assert!(
        !rendered.contains(TOKEN),
        "credential leaked into Debug output: {rendered}"
    );
    assert!(rendered.contains("<redacted>"), "got: {rendered}");
    assert!(rendered.contains("telemetry-collector.example.workers.dev"));
}

#[test]
fn default_tenant_is_sent_only_when_configured() {
    let without = backend().metrics_request(&snapshot()).unwrap().unwrap();
    assert_eq!(header(&without, TENANT_HEADER), None);

    let with = CloudflareBackend::new(ENDPOINT, TOKEN)
        .with_default_tenant(Some("acme".to_string()))
        .metrics_request(&snapshot())
        .unwrap()
        .unwrap();
    assert_eq!(header(&with, TENANT_HEADER), Some("acme"));
}

#[test]
fn blank_default_tenant_is_treated_as_unset() {
    let backend =
        CloudflareBackend::new(ENDPOINT, TOKEN).with_default_tenant(Some("  ".to_string()));
    assert_eq!(backend.default_tenant(), None);
}

#[test]
fn empty_batches_produce_no_request() {
    let backend = backend();
    assert!(backend.traces_request("ferrogate", &[]).unwrap().is_none());
    assert!(backend.logs_request("ferrogate", &[]).unwrap().is_none());
}

#[test]
fn unsupported_signals_are_skipped() {
    let backend = CloudflareBackend::new(ENDPOINT, TOKEN)
        .with_signals(vec![ObservabilitySignal::Trace, ObservabilitySignal::Log]);
    assert!(backend.metrics_request(&snapshot()).unwrap().is_none());
    assert!(backend
        .traces_request("ferrogate", &[span()])
        .unwrap()
        .is_some());
}

#[test]
fn validate_accepts_an_https_collector() {
    assert!(backend().validate().is_ok());
}

#[test]
fn validate_refuses_plaintext_to_a_remote_collector() {
    let error = CloudflareBackend::new("http://collector.example.com", TOKEN)
        .validate()
        .unwrap_err();
    assert!(
        matches!(error, ObservabilityConfigError::InsecureEndpoint { .. }),
        "got: {error:?}"
    );
}

#[test]
fn validate_allows_plaintext_loopback_for_wrangler_dev() {
    for endpoint in [
        "http://localhost:8787",
        "http://127.0.0.1:8787",
        "http://[::1]:8787",
        "http://localhost:8787/ingest",
    ] {
        assert!(
            CloudflareBackend::new(endpoint, TOKEN).validate().is_ok(),
            "expected {endpoint} to be accepted"
        );
    }
}

#[test]
fn validate_refuses_a_host_that_merely_looks_like_loopback() {
    for endpoint in [
        "http://localhost.evil.com",
        "http://127.0.0.1.evil.com",
        "http://user@evil.com",
        "http://evil.com/localhost",
    ] {
        let error = CloudflareBackend::new(endpoint, TOKEN)
            .validate()
            .unwrap_err();
        assert!(
            matches!(error, ObservabilityConfigError::InsecureEndpoint { .. }),
            "expected {endpoint} to be refused, got: {error:?}"
        );
    }
}

#[test]
fn validate_requires_a_credential() {
    let error = CloudflareBackend::new(ENDPOINT, "   ")
        .validate()
        .unwrap_err();
    assert!(matches!(
        error,
        ObservabilityConfigError::MissingCredential { .. }
    ));
}

#[test]
fn validate_refuses_a_credential_that_could_split_the_request() {
    let error = CloudflareBackend::new(ENDPOINT, "token\r\nX-Injected: 1")
        .validate()
        .unwrap_err();
    assert!(
        matches!(error, ObservabilityConfigError::InvalidCredential { .. }),
        "got: {error:?}"
    );
}

#[test]
fn validate_still_checks_the_endpoint_shape() {
    let error = CloudflareBackend::new("", TOKEN).validate().unwrap_err();
    assert!(matches!(
        error,
        ObservabilityConfigError::MissingEndpoint { .. }
    ));
}

File: packages/observability/test/cloudflare.test.ts
import { describe, expect, test } from "vitest";
import {
  CloudflareBackend,
  defaultGatewayMetricsSnapshot,
  endpointProtectsCredentials,
  ObservabilitySignal,
  otlpAttribute,
  TENANT_HEADER,
  type GatewayMetricsSnapshot,
  type OtlpHttpRequest,
  type OtlpLogRecord,
  type OtlpSpanRecord,
} from "../src/index.js";

const TOKEN = "s3cr3t-collector-token";
const ENDPOINT = "https://telemetry-collector.example.workers.dev";

function backend(): CloudflareBackend {
  return new CloudflareBackend(ENDPOINT, TOKEN);
}

function snapshot(): GatewayMetricsSnapshot {
  return { ...defaultGatewayMetricsSnapshot(), serviceName: "ferrogate" };
}

function span(): OtlpSpanRecord {
  return {
    traceId: "0af7651916cd43dd8448eb211c80319c",
    spanId: "b7ad6b7169203331",
    name: "ferrogate.gateway.request",
    startTimeUnixNano: 1,
    endTimeUnixNano: 2,
    attributes: [otlpAttribute("tenant", "acme")],
  };
}

function log(): OtlpLogRecord {
  return { severityText: "INFO", body: "request", timeUnixNano: 1, attributes: [] };
}

function header(request: OtlpHttpRequest, name: string): string | undefined {
  return request.headers.find(
    ([key]) => key.toLowerCase() === name.toLowerCase(),
  )?.[1];
}

describe("CloudflareBackend", () => {
  test("every signal carries the bearer credential", () => {
    const b = backend();
    const requests = [
      b.metricsRequest(snapshot()),
      b.tracesRequest("ferrogate", [span()]),
      b.logsRequest("ferrogate", [log()]),
    ];
    for (const request of requests) {
      expect(request).not.toBeNull();
      expect(header(request as OtlpHttpRequest, "authorization")).toBe(
        `Bearer ${TOKEN}`,
      );
    }
  });

  test("signals target the standard OTLP paths", () => {
    const b = backend();
    expect(b.metricsRequest(snapshot())?.url).toBe(`${ENDPOINT}/v1/metrics`);
    expect(b.tracesRequest("ferrogate", [span()])?.url).toBe(
      `${ENDPOINT}/v1/traces`,
    );
    expect(b.logsRequest("ferrogate", [log()])?.url).toBe(`${ENDPOINT}/v1/logs`);
  });

  test("debug output redacts the credential", () => {
    const rendered = backend().redactedDebug();
    expect(rendered).not.toContain(TOKEN);
    expect(rendered).toContain("<redacted>");
    expect(rendered).toContain("telemetry-collector.example.workers.dev");
  });

  test("default tenant is sent only when configured", () => {
    const without = backend().metricsRequest(snapshot()) as OtlpHttpRequest;
    expect(header(without, TENANT_HEADER)).toBeUndefined();

    const withTenant = new CloudflareBackend(ENDPOINT, TOKEN)
      .withDefaultTenant("acme")
      .metricsRequest(snapshot()) as OtlpHttpRequest;
    expect(header(withTenant, TENANT_HEADER)).toBe("acme");
  });

  test("blank default tenant is treated as unset", () => {
    const b = new CloudflareBackend(ENDPOINT, TOKEN).withDefaultTenant("  ");
    expect(b.defaultTenant()).toBeUndefined();
  });

  test("empty batches produce no request", () => {
    const b = backend();
    expect(b.tracesRequest("ferrogate", [])).toBeNull();
    expect(b.logsRequest("ferrogate", [])).toBeNull();
  });

  test("unsupported signals are skipped", () => {
    const b = new CloudflareBackend(ENDPOINT, TOKEN).withSignals([
      ObservabilitySignal.Trace,
      ObservabilitySignal.Log,
    ]);
    expect(b.metricsRequest(snapshot())).toBeNull();
    expect(b.tracesRequest("ferrogate", [span()])).not.toBeNull();
  });

  test("validate accepts an https collector", () => {
    expect(backend().validate()).toBeNull();
  });

  test("validate refuses plaintext to a remote collector", () => {
    expect(
      new CloudflareBackend("http://collector.example.com", TOKEN).validate()
        ?.errorKind,
    ).toBe("InsecureEndpoint");
  });

  test("validate allows plaintext loopback for wrangler dev", () => {
    for (const endpoint of [
      "http://localhost:8787",
      "http://127.0.0.1:8787",
      "http://[::1]:8787",
      "http://localhost:8787/ingest",
    ]) {
      expect(new CloudflareBackend(endpoint, TOKEN).validate()).toBeNull();
    }
  });

  test("validate refuses a host that merely looks like loopback", () => {
    for (const endpoint of [
      "http://localhost.evil.com",
      "http://127.0.0.1.evil.com",
      "http://user@evil.com",
      "http://evil.com/localhost",
    ]) {
      expect(
        new CloudflareBackend(endpoint, TOKEN).validate()?.errorKind,
      ).toBe("InsecureEndpoint");
    }
  });

  test("validate requires a credential", () => {
    expect(
      new CloudflareBackend(ENDPOINT, "   ").validate()?.errorKind,
    ).toBe("MissingCredential");
  });

  test("validate refuses a credential that could split the request", () => {
    expect(
      new CloudflareBackend(ENDPOINT, "token\r\nX-Injected: 1").validate()
        ?.errorKind,
    ).toBe("InvalidCredential");
  });

  test("validate still checks the endpoint shape", () => {
    expect(new CloudflareBackend("", TOKEN).validate()?.errorKind).toBe(
      "MissingEndpoint",
    );
  });
});

describe("endpointProtectsCredentials", () => {
  test("https always protects; loopback http protects; remote http does not", () => {
    expect(endpointProtectsCredentials("https://anything")).toBe(true);
    expect(endpointProtectsCredentials("http://localhost")).toBe(true);
    expect(endpointProtectsCredentials("http://[::1]:9000/x")).toBe(true);
    expect(endpointProtectsCredentials("http://remote.example.com")).toBe(false);
    expect(endpointProtectsCredentials("ftp://localhost")).toBe(false);
  });
});
